api/models/metadata.go

```diff
@@ -0,0 +1,230 @@
+package models
+
+import (
+	"encoding/json"
+	"fmt"
+	"bytes"
+	"database/sql/driver"
+	"errors"
+)
+
+// Metadata encapsulates key-value metadata. The structure is immutable via its public API and nil-safe for its contract
+// permissive nilability is here to simplify updates and reduce the need for nil handling in extensions - metadata should be updated by over-writing the original object:
+//  oldMd := EmptyMetadata()
+//  newMd := oldMd.With("fooKey",1)
+//  // old MD remains empty
+// Metadata is lenable
+type Metadata map[string]*metadataValue
```

Comment:
slightly curious whether doing `map[string]interface{}` makes our lives any easier, it seems essentially to be what we want but with some logic to validate key / value sizes on top of that. if so, I assume it'd mean since we always ingest these from json through the http api (for now), we simply need to implement the key / value validations inside of the `UnmarshalJSON` logic, and this would give us slightly more control over that validation (at present, only values have a custom UnmarshalJSON, which reads the whole thing in before validation). haven't played with this concretely, I think the keys are easy (seek until `"` or n > maxKeyBytes), but the values we'd need to wrap a limit bytes reader around and unmarshal an `interface{}`, which is perhaps a little easier than reading into a bytes buffer.

as it stands, the code here to validate is vulnerable to e.g. a user sending in a 5GB key or value will allocate that space (short of any http level protections, which exist, but still likely exceed the maximums imposed here) before we validate that it is too large. I am not sure if this is an explicit goal to block this behavior.

up to you, I think it'd be possible to play with later since this is essentially `map[string]interface{}` we can change the implementation but the data would stay the same.

Reply:
Protecting the heap wasn't wasn't the goal here, the the size validation here is more to protect the DB.  Given we aren't taking unbounded input data anywhere (I think)  I'm pretty sure a reasonable cap on ingress size in middleware and not making request size validation a concern of the business logic is the right way to go here. 

For read I want extensions  consuming metadata  to be able to bind an object/array to whatever they want (either naive map[string]interface{} or to their own structs/slices) - so forcing the unmarshalled format to the former seemed inconvenience (people end up marshalling and then unmarshalling again) . 

For write I wanted to apply validation at the point of setting  including value length checks (i.e. its an error in the extension if it sets an invalid MD key) , which entails marshalling so it seemed to make sense to marshal early here.

Reply:
> For read I want extensions consuming metadata to be able to bind an object/array to whatever they want (either naive map[string]interface{} or to their own structs/slices) - so forcing the unmarshalled format to the former seemed inconvenience (people end up marshalling and then unmarshalling again) .

that is a good point. in that case `interface{}` (no map) seems easiest, even, but it's weird to work with for storage / json. bytes seems a fine decision.